test/shared/src/main/scala2/zio/test/Macros.scala

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import scala.reflect.macros.blackbox.Context
+
+private[test] object Macros {
+
+  def assertCompiles_impl(c: Context)(code: c.Expr[String]): c.Expr[TestResult] = {
+    import c.universe._
+    try {
+      c.typecheck(c.parse(c.eval(c.Expr[String](c.untypecheck(code.tree)))))
+      c.Expr(q"assert(true, Assertion.isTrue)")
+    } catch {
+      case _: Throwable => c.Expr(q"assert(false, Assertion.isTrue)")
```

Comment:
Is it possible to splice the error message here?

Reply:
I think it is. The reason I didn't is that on Dotty `typeChecks` just returns whether or not code compiles. So if we want to have a consistent API across platforms we would have to do something a little fancier on Dotty to get the error messages there as well. While I haven't looked into it I assume that is doable. Is that a direction we want to go?

Reply:
@adamgfraser 
I think No, because the staging/quote/tasty APIs are not stable yet, so we'd have to fix them when upgrading version.
However, nearly all users are going to use this API from Scala 2, in which case they'd probably find an error message in their test very helpful, as otherwise on a test failure they'd have to move the failing code outside of the string to get the error message.

So, my opinion is that we should leave dotty as is because the relevant APIs aren't stable and there are very few users who'll notice, but it would be pretty helpful for our users to get the error messages on Scala 2.

Reply:
@neko-kai I think those are excellent points. Let me work on a follow up PR to add that functionality.

Reply:
@adamgfraser Thanks!